### PR TITLE
Implement Nearest Friendly Evaluation Function

### LINKED ISSUE #295: Implement nearest_friendly Evaluation Function
### Description of the task

Implement the nearest_friendly evaluation function.

### Outline the task in more detail

We would like to implement the nearest_friendly evaluation function in our new codebase. It is currently defined in the old codebase at `ai/hl/stp/evaluation/team.cpp`. With respect to the interface of this function (ie. what it takes and received), there should be similar types available in the new codebase. That said, it is possible that this function takes a type that is larger then it needs (ex. taking a `World` when it just needs a `Team` and a `ball`). Basically, try to use the smallest type you possible can; this will make it easier to test, review, develop, and work with in the future. 

### Acceptance criteria

- [x] Function nearest_friendly is implemented
- [x] Function nearest_friendly is unit tested

### PR DESCRIPTION
**Pull request by matthewberends**
### Description

Implemented the nearest friendly evaluation function.

### Testing Done

Wrote unit tests and project builds successfully.

### Resolved Issues

resolves #295 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [x] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom`
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/thunderbots/software/test/evaluation/team.cpp [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+   protected:
+    void SetUp() override
+    {
+        current_time = Timestamp::fromSeconds(123);
+    }
+
+    Timestamp current_time;
+};
+
+TEST_F(EvaluationTeamTest, one_robot)
```
Comment (garethellis0) [inline src/thunderbots/software/test/evaluation/team.cpp:19]: Could we add a test for zero robots?

File: src/thunderbots/software/ai/hl/stp/evaluation/team.h [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+     * Finds the robot on the friendly team that is closest to some reference point.
+     *
+     * @param friendly_team
+     * @param ref_point The point where the distance to each robot will be measured.
+     *
+     * @return Robot that is closest to the reference point.
+     */
+    Robot nearest_friendly(const Team friendly_team, const Point ref_point);
+
+};  // namespace Evaluation
```
Comment (garethellis0) [inline src/thunderbots/software/ai/hl/stp/evaluation/team.h]: nit; github complaining about end-of-file

File: src/thunderbots/software/ai/hl/stp/evaluation/team.cpp [hunk 3]
```diff
@@ -0,0 +1,20 @@
+#include "ai/hl/stp/evaluation/team.h"
+
+
+Robot Evaluation::nearest_friendly(const Team friendly_team, const Point ref_point)
+{
+    unsigned nearestRobotId = friendly_team.getAllRobots()[0].id();
```
Comment (garethellis0) [inline src/thunderbots/software/ai/hl/stp/evaluation/team.cpp:12]: hmm, what if the team is empty? I'm genuinely unsure of what the best way to handle this case would be. The two options I can think of would be to return an optional (https://en.cppreference.com/w/cpp/utility/optional) or throw an exception. @MathewMacDougall what do you think? We're gonna see this come up a lot, so it might be good to sit down with you and @matthewberends and figure out what we want to do in these cases?

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Good stuff, pretty clean, nice tests.

File: src/thunderbots/software/ai/hl/stp/evaluation/team.cpp (hunk 3 above)
Comment (matthewberends) [reply src/thunderbots/software/ai/hl/stp/evaluation/team.cpp:12]: Lots of the functions in Team class seem to use optionals (ex: getRobotById) so having a team with no robots seems like a valid state. With that in mind I think we should return an optional for this evaluation function.

Comment (garethellis0) [reply src/thunderbots/software/ai/hl/stp/evaluation/team.cpp:12]: Ah, nice job looking into that! Sounds good to me!

File: src/thunderbots/software/test/evaluation/team.cpp (hunk 1 above)
Comment (matthewberends) [reply src/thunderbots/software/test/evaluation/team.cpp:19]: Done

File: src/thunderbots/software/ai/hl/stp/evaluation/team.h (hunk 2 above)
Comment (matthewberends) [reply src/thunderbots/software/ai/hl/stp/evaluation/team.h]: Fixed

File: src/thunderbots/software/ai/hl/stp/evaluation/team.cpp (hunk 3 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/hl/stp/evaluation/team.cpp:12]: I agree with using optional here :+1:

File: src/thunderbots/software/test/evaluation/team.cpp [hunk 4]
```diff
@@ -0,0 +1,120 @@
+
+#include "ai/hl/stp/evaluation/team.h"
+
+#include <gtest/gtest.h>
+
+// using namespace Evaluation;
```
Comment (MathewMacDougall) [inline src/thunderbots/software/test/evaluation/team.cpp:6]: Might as well remove this line if it's going to be commented out.

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
LGTM pending the change to handle the case of empty teams and return optionals.

**Review by garethellis0 — APPROVED** [review]
Great job getting these changes in! LGTM :+1: :tada:

**Review by MathewMacDougall — APPROVED** [review]
LGTM :+1:. Nice job with the optionals.
